tools/subtype-gen/README.md

```diff
@@ -0,0 +1,320 @@
+# Rules
+
+A rules file contains zero or more rules.
+A single rule consist of the two properties: The super-type (`super`) for which this rule applies to,
+and a predicate (`predicate`) defining the conditions that needs to be satisfied by a type to be a subtype of this super-type.
+
+```yaml
+rules :
+  - super: T1
+    predicate: P1
+
+  - super: T2
+    predicate: P2
+    
+  ...
+```
+
+## Predicates
+
+A predicate is a condition that would become either `true` or `false` depending on the inputs.
+Below are the different types of predicates that can be used to define the subtype rules for a given super-type.
+
+### Always Predicate (`always`)
+ 
+Represents a condition that is always true.
+
+### Never Predicate (`never`)
+
+Represents a condition that is always false.
+
+### Not Predicate (`not`)
+
+Negates a nested predicate.
+
+### And Predicate (`and`)
+
+Represent a list of nested predicate that needs to be satisfied.
+An `and` predicate becomes true only if all of its nested predicates are also true.
+e.g:
+
+```yaml
+and:
+  - P1
+  - P2
+```
+
+### Or Predicate (`or`)
+
+Represent a list of nested predicate that could be satisfied.
+An `or` predicate becomes true if at-least one of its nested predicates are also true.
+e.g:
+
+```yaml
+and:
+  - P1
+  - P2
+```
+
+### Equals Predicate (`equals`)
```

Comment:
Maybe note here that this corresponds to Go `==`, not e.g. an `Equals` method call